pkg/csi/service/wcp/controller.go

```diff
@@ -1237,149 +1239,228 @@ func (c *controller) ControllerUnpublishVolume(ctx context.Context, req *csi.Con
+					vCenterHost = key
+					dcMorefValue = value
+				}
+				vc, err := c.manager.VcenterManager.GetVirtualCenter(ctx, vCenterHost)
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"cannot get virtual center %s from virtualcentermanager while attaching disk with error %+v",
+						vc.Config.Host, err)
+				}
+				// Connect to VC.
+				err = vc.Connect(ctx)
+				if err != nil {
+					return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
+						"failed to connect to Virtual Center: %s", vc.Config.Host)
+				}
+				podVM, err := getVMByInstanceUUIDInDatacenter(ctx, vc, dcMorefValue, vmuuid)
+				if err != nil {
+					if err != cnsvsphere.ErrVMNotFound {
 						return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
-							"cannot get virtual center %s from virtualcentermanager while attaching disk with error %+v",
-							vc.Config.Host, err)
+							"failed to get the PodVM Moref from the PodVM UUID: %s in datacenter: %s with err: %+v",
+							vmuuid, dcMorefValue, err)
 					}
-					// Connect to VC.
-					err = vc.Connect(ctx)
+					log.Infof("virtual machine not found for vmUUID %q. "+
+						"Thus, assuming the volume is detached.", vmuuid)
+				}
+
+				if podVM != nil {
+					faultType, err := common.DetachVolumeUtil(ctx, c.manager.VolumeManager, podVM, req.VolumeId)
 					if err != nil {
-						return nil, csifault.CSIInternalFault, logger.LogNewErrorCodef(log, codes.Internal,
-							"failed to connect to Virtual Center: %s", vc.Config.Host)
+						return nil, faultType, logger.LogNewErrorCodef(log, codes.Internal,
+							"failed to detach disk: %+q from node: %q err %+v", req.VolumeId, req.NodeId, err)
 					}
-					isStillAttached := false
-					timeout := 4 * time.Minute
-					pollTime := time.Duration(5) * time.Second
-					var podVM *cnsvsphere.VirtualMachine
-					err = wait.PollUntilContextTimeout(ctx, pollTime, timeout, true,
-						func(ctx context.Context) (bool, error) {
-							podVM, err = getVMByInstanceUUIDInDatacenter(ctx, vc, dcMorefValue, v)
-							if err != nil {
-								if err == cnsvsphere.ErrVMNotFound {
-									log.Infof("virtual machine not found for vmUUID %q. "+
-										"Thus, assuming the volume is detached.", v)
-									return true, err
-								}
-								if err == cnsvsphere.ErrInvalidVC {
+				}
+			}
+
+			if commonco.ContainerOrchestratorUtility.IsFSSEnabled(ctx, common.FakeAttach) {
+				// Check if the volume was fake attached and unmark it as not fake
+				// attached.
+				if err := commonco.ContainerOrchestratorUtility.ClearFakeAttached(ctx, req.VolumeId); err != nil {
+					msg := fmt.Sprintf("Failed to unmark volume as not fake attached. Error: %v", err)
+					log.Error(msg)
+					return nil, csifault.CSIInternalFault, err
+				}
+			}
+			return &csi.ControllerUnpublishVolumeResponse{}, "", nil
+		} else {
+			var isVADeleted bool
```

Comment:
nit: Since it's the same logic we're invoking here and the case when WCP FSS is enabled, can we wrap getting VA object in a helper method and use that at both places?

Reply:
Since it was only ~10 lines, I ended up keeping it as is.